dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java

```diff
@@ -333,8 +334,10 @@ public TypedFormula<BooleanType, BooleanFormula> visitBoolUnaryExpression(BoolUn
             if (expr.isNoop()) {
                 return inner;
             } else if (context.useIntegers) {
-                //TODO If narrowing, constrain the value (mod 2^bitWidth).
-                return new TypedFormula<>(expr.getType(), inner.formula());
+                final BigInteger highValue = BigInteger.TWO.pow(types.getMemorySizeInBits(expr.getType()));
+                final IntegerFormulaManager imgr = integerFormulaManager();
+                final IntegerFormula truncated = imgr.modulo((IntegerFormula) inner.formula(), imgr.makeNumber(highValue));
+                return new TypedFormula<>(expr.getType(), truncated);
```

Comment:
This is only necessary for the truncation case. The cast may also be an extension.